[Bug]: read tool path guard rejects valid /agent/ mount paths in workspaceAccess: "ro" sandboxed sessions

### Bug type

Crash (process/app exits or hangs)

### Summary

When workspaceAccess is set to "ro", the agent workspace is mounted read-only at /agent/ inside the  container. However, the gateway-side wrapSandboxPathGuard in pi-tools.read.ts does not map /agent/  paths back to the host workspace root, causing all read tool calls with container-internal paths to fail with "Path escapes sandbox root".                                                                 
This is the same class of bug as #9560 (which fixed /workspace/ path mapping for containerWorkdir), but for the workspaceAccess: "ro" mount point.                

### Steps to reproduce

1. Configure sandbox with workspaceAccess: "ro":
  {
    "agents": {
      "defaults": {
        "sandbox": {
          "mode": "non-main",
          "workspaceAccess": "ro"
        }
      }
    }
  }
  2. Spawn a subagent (which runs in the sandbox container)
  3. Subagent attempts to read a file using the read tool with the container mount path:
  read /agent/some-project/src/main.py

### Expected behavior

The read tool should map /agent/... to the host workspace path and succeed (respecting read-only access).

### Actual behavior

[ERROR] [tools] read failed: Path escapes sandbox root
  (~/.openclaw/sandboxes/agent-myagent-subagent-abc123--def456): /agent/some-project/src/main.py

  The path guard compares /agent/... directly against the sandbox root (~/.openclaw/sandboxes/...), finds
   no common prefix, and rejects it.

  Additional error variants observed:

  The AI also tries alternative paths to reach the files, all rejected:
  read failed: Path escapes sandbox root: ~/workspace/skills/my-skill/SKILL.md
  read failed: Sandbox FS error (ENOENT):
  .../sandboxes/agent-myagent-subagent-abc123/workspace/skills/my-skill/SKILL.md
  read failed: Path escapes sandbox root: /tmp/output/results.json

### OpenClaw version

2026.3.2

### Operating system

macOS

### Install method

npm global

### Logs, screenshots, and evidence

```shell

```

### Impact and severity

Read tool broken in ro sandbox mode

### Additional information

The #9560 fix (commit 73fab7e44) added containerWorkdir mapping to wrapSandboxPathGuard, but only for
  docker.workdir paths (default /workspace/). When workspaceAccess: "ro", the mount point is /agent/, not
   /workspace/, so the mapping doesn't apply.

  The path guard needs to also map the ro mount path (/agent/) to the host workspace root, similar to how
   it maps containerWorkdir.

Suggested Fix

  Extend the containerWorkdir mapping in wrapSandboxPathGuard to also cover the workspaceAccess: "ro"
  mount point (/agent/). The mount path is deterministic based on the workspaceAccess setting:
  - "ro" → /agent/
  - "rw" → /workspace/
  - "none" → N/A (no workspace mount)

  Additionally, consider mapping /tmp/ paths for container-local reads, since sandboxed sessions write to
   /tmp/ but can't read them back with the read tool.

  Workaround

  Instruct the AI to use exec cat /agent/path/to/file instead of read /agent/path/to/file. The exec tool
  runs inside the container where /agent/ is a real mount point, bypassing the gateway-side path guard.
  However, this wastes significant tokens as the AI repeatedly attempts read before falling back to exec.

  Related

  - #9560 — Sandbox path validation fails for container-internal paths (/workspace/...) — fixed for
  containerWorkdir only

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -67,7 +67,7 @@ Docs: https://docs.openclaw.ai
 - Codex app-server: refresh Codex account rate limits after subscription usage-limit failures so Discord and other channel replies can show the next reset time instead of saying Codex returned none. Thanks @pashpashpash.
 - Tasks: route group and channel task completions through the requester session so the parent agent can send the visible summary instead of stopping at a generic task-status line. Fixes #77251. (#77365) Thanks @funmerlin.
 - Telegram: preserve blank lines between manually indented bullet blocks and following numbered sections in rendered replies. Fixes #76998. Thanks @evgyur.
-- Agents/sandbox: allow read-only sandbox sessions to read the `/agent` workspace mount while keeping write/edit/apply_patch workspace-only guarded, restoring `read /agent/...` for `workspaceAccess: "ro"`. Fixes #39497. Thanks @teosborne.
+- Agents/sandbox: allow read-only sandbox sessions to read the `/agent` workspace mount while keeping write/edit/apply_patch workspace-only guarded, restoring `read /agent/...` for `workspaceAccess: "ro"`. Fixes #39497. Thanks @stainlu and @teosborne.
 - Slack: pass configured agent identity through draft preview sends so partial streaming replies keep custom username/avatar on the initial Slack message. Fixes #38235. (#38237) Thanks @lacymorrow.
 - Slack: support `allowBots: "mentions"` for bot-authored messages that mention the receiving bot, matching the documented Discord-style mode without accepting every bot message. Fixes #43587. (#43588) Thanks @raw34.
 - Slack: refresh private file URLs with `files.info` when inbound DM file events omit or stale attachment URLs, preventing file attachments from being dropped before media hydration. Fixes #50129. (#50200) Thanks @smartchainark.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -67,6 +67,7 @@ Docs: https://docs.openclaw.ai
 - Codex app-server: refresh Codex account rate limits after subscription usage-limit failures so Discord and other channel replies can show the next reset time instead of saying Codex returned none. Thanks @pashpashpash.
 - Tasks: route group and channel task completions through the requester session so the parent agent can send the visible summary instead of stopping at a generic task-status line. Fixes #77251. (#77365) Thanks @funmerlin.
 - Telegram: preserve blank lines between manually indented bullet blocks and following numbered sections in rendered replies. Fixes #76998. Thanks @evgyur.
+- Agents/sandbox: allow read-only sandbox sessions to read the `/agent` workspace mount while keeping write/edit/apply_patch workspace-only guarded, restoring `read /agent/...` for `workspaceAccess: "ro"`. Fixes #39497. Thanks @teosborne.
 - Slack: pass configured agent identity through draft preview sends so partial streaming replies keep custom username/avatar on the initial Slack message. Fixes #38235. (#38237) Thanks @lacymorrow.
 - Slack: support `allowBots: "mentions"` for bot-authored messages that mention the receiving bot, matching the documented Discord-style mode without accepting every bot message. Fixes #43587. (#43588) Thanks @raw34.
 - Slack: refresh private file URLs with `files.info` when inbound DM file events omit or stale attachment URLs, preventing file attachments from being dropped before media hydration. Fixes #50129. (#50200) Thanks @smartchainark.

--- a/src/agents/pi-tools.read.ts
+++ b/src/agents/pi-tools.read.ts
@@ -351,19 +351,15 @@ function mapContainerPathToWorkspaceRoot(params: {
   root: string;
   containerWorkdir?: string;
 }): string {
-  const containerWorkdir = params.containerWorkdir?.trim();
-  if (!containerWorkdir) {
-    return params.filePath;
-  }
-  const normalizedWorkdir = containerWorkdir.replace(/\\/g, "/").replace(/\/+$/, "");
-  if (!normalizedWorkdir.startsWith("/")) {
-    return params.filePath;
-  }
-  if (!normalizedWorkdir) {
-    return params.filePath;
-  }
+  return mapContainerPathToRoot({
+    filePath: params.filePath,
+    root: params.root,
+    containerRoot: params.containerWorkdir,
+  }).filePath;
+}
 
-  let candidate = params.filePath.startsWith("@") ? params.filePath.slice(1) : params.filePath;
+function resolveContainerPathCandidate(filePath: string): string | null {
+  let candidate = filePath.startsWith("@") ? filePath.slice(1) : filePath;
   if (/^file:\/\//i.test(candidate)) {
     const localFilePath = trySafeFileURLToPath(candidate);
     if (localFilePath) {
@@ -375,47 +371,65 @@ function mapContainerPathToWorkspaceRoot(params: {
       try {
         parsed = new URL(candidate);
       } catch {
-        return params.filePath;
+        return filePath;
       }
       if (parsed.protocol !== "file:") {
-        return params.filePath;
+        return filePath;
       }
       const host = parsed.hostname.trim().toLowerCase();
       if (host && host !== "localhost") {
-        return params.filePath;
+        return filePath;
       }
       if (hasEncodedFileUrlSeparator(parsed.pathname)) {
-        return params.filePath;
+        return filePath;
       }
       let normalizedPathname: string;
       try {
         normalizedPathname = decodeURIComponent(parsed.pathname).replace(/\\/g, "/");
       } catch {
-        return params.filePath;
-      }
-      if (
-        normalizedPathname !== normalizedWorkdir &&
-        !normalizedPathname.startsWith(`${normalizedWorkdir}/`)
-      ) {
-        return params.filePath;
+        return filePath;
       }
       candidate = normalizedPathname;
     }
   }
+  return candidate;
+}
+
+function mapContainerPathToRoot(params: {
+  filePath: string;
+  root: string;
+  containerRoot?: string;
+}): { filePath: string; matched: boolean } {
+  const containerRoot = params.containerRoot?.trim();
+  if (!containerRoot) {
+    return { filePath: params.filePath, matched: false };
+  }
+  const normalizedRoot = containerRoot.replace(/\\/g, "/").replace(/\/+$/, "");
+  if (!normalizedRoot.startsWith("/") || !normalizedRoot) {
+    return { filePath: params.filePath, matched: false };
+  }
+
+  const candidate = resolveContainerPathCandidate(params.filePath);
+  if (candidate === null) {
+    return { filePath: params.filePath, matched: false };
+  }
 
   const normalizedCandidate = candidate.replace(/\\/g, "/");
-  if (normalizedCandidate === normalizedWorkdir) {
-    return path.resolve(params.root);
+  if (normalizedCandidate === normalizedRoot) {
+    return { filePath: path.resolve(params.root), matched: true };
   }
-  const prefix = `${normalizedWorkdir}/`;
+  const prefix = `${normalizedRoot}/`;
   if (!normalizedCandidate.startsWith(prefix)) {
-    return candidate;
+    return { filePath: candidate, matched: false };
   }
   const relative = normalizedCandidate.slice(prefix.length);
   if (!relative) {
-    return path.resolve(params.root);
+    return { filePath: path.resolve(params.root), matched: true };
   }
-  return path.resolve(params.root, ...relative.split("/").filter(Boolean));
+  return {
+    filePath: path.resolve(params.root, ...relative.split("/").filter(Boolean)),
+    matched: true,
+  };
 }
 
 export function resolveToolPathAgainstWorkspaceRoot(params: {
@@ -576,6 +590,10 @@ export function wrapToolWorkspaceRootGuardWithOptions(
   tool: AnyAgentTool,
   root: string,
   options?: {
+    additionalContainerMounts?: readonly {
+      containerRoot: string;
+      hostRoot: string;
+    }[];
     containerWorkdir?: string;
     pathParamKeys?: readonly string[];
     normalizeGuardedPathParams?: boolean;
@@ -593,12 +611,32 @@ export function wrapToolWorkspaceRootGuardWithOptions(
         if (typeof filePath !== "string" || !filePath.trim()) {
           continue;
         }
-        const sandboxPath = mapContainerPathToWorkspaceRoot({
+        let guardedRoot = root;
+        const workspaceMapping = mapContainerPathToRoot({
           filePath,
           root,
-          containerWorkdir: options?.containerWorkdir,
+          containerRoot: options?.containerWorkdir,
         });
-        const sandboxResult = await assertSandboxPath({ filePath: sandboxPath, cwd: root, root });
+        let sandboxPath = workspaceMapping.filePath;
+        if (!workspaceMapping.matched) {
+          for (const mount of options?.additionalContainerMounts ?? []) {
+            const mountMapping = mapContainerPathToRoot({
+              filePath,
+              root: mount.hostRoot,
+              containerRoot: mount.containerRoot,
+            });
+            if (mountMapping.matched) {
+              guardedRoot = path.resolve(mount.hostRoot);
+              sandboxPath = mountMapping.filePath;
+              break;
+            }
+          }
+        }
+        const sandboxResult = await assertSandboxPath({
+          filePath: sandboxPath,
+          cwd: guardedRoot,
+          root: guardedRoot,
+        });
         if (options?.normalizeGuardedPathParams && record) {
           normalizedRecord ??= { ...record };
           normalizedRecord[key] = sandboxResult.resolved;

--- a/src/agents/pi-tools.read.workspace-root-guard.test.ts
+++ b/src/agents/pi-tools.read.workspace-root-guard.test.ts
@@ -173,6 +173,40 @@ describe("wrapToolWorkspaceRootGuardWithOptions", () => {
     });
   });
 
+  it("maps additional container mounts to their own guarded host roots", async () => {
+    const { tool } = createToolHarness();
+    const agentRoot = "/tmp/agent-root";
+    const wrapped = wrapToolWorkspaceRootGuardWithOptions(tool, root, {
+      additionalContainerMounts: [{ containerRoot: "/agent", hostRoot: agentRoot }],
+      containerWorkdir: "/workspace",
+    });
+
+    await wrapped.execute("tc-agent-mount", { path: "/agent/docs/readme.md" });
+
+    expect(mocks.assertSandboxPath).toHaveBeenCalledWith({
+      filePath: path.resolve(agentRoot, "docs", "readme.md"),
+      cwd: agentRoot,
+      root: agentRoot,
+    });
+  });
+
+  it("maps file URLs under additional container mounts", async () => {
+    const { tool } = createToolHarness();
+    const agentRoot = "/tmp/agent-root";
+    const wrapped = wrapToolWorkspaceRootGuardWithOptions(tool, root, {
+      additionalContainerMounts: [{ containerRoot: "/agent", hostRoot: agentRoot }],
+      containerWorkdir: "/workspace",
+    });
+
+    await wrapped.execute("tc-agent-file-url", { path: "file:///agent/docs/readme.md" });
+
+    expect(mocks.assertSandboxPath).toHaveBeenCalledWith({
+      filePath: path.resolve(agentRoot, "docs", "readme.md"),
+      cwd: agentRoot,
+      root: agentRoot,
+    });
+  });
+
   it("does not guard outPath by default", async () => {
     const { tool } = createToolHarness();
     const wrapped = wrapToolWorkspaceRootGuardWithOptions(tool, root, {

--- a/src/agents/pi-tools.sandbox-mounted-paths.workspace-only.test.ts
+++ b/src/agents/pi-tools.sandbox-mounted-paths.workspace-only.test.ts
@@ -9,6 +9,7 @@ import {
   createSandboxedWriteTool,
   wrapToolWorkspaceRootGuardWithOptions,
 } from "./pi-tools.read.js";
+import { SANDBOX_AGENT_WORKSPACE_MOUNT } from "./sandbox/constants.js";
 import {
   expectReadWriteEditTools,
   expectReadWriteTools,
@@ -68,6 +69,15 @@ function createSandboxFsTools(params: { sandbox: UnsafeMountedSandbox; workspace
   }
   return tools.map((tool) =>
     wrapToolWorkspaceRootGuardWithOptions(tool, params.sandbox.workspaceDir, {
+      additionalContainerMounts:
+        tool.name === "read" && params.sandbox.workspaceAccess === "ro"
+          ? [
+              {
+                containerRoot: SANDBOX_AGENT_WORKSPACE_MOUNT,
+                hostRoot: params.sandbox.agentWorkspaceDir,
+              },
+            ]
+          : undefined,
       containerWorkdir: params.sandbox.containerWorkdir,
     }),
   );
@@ -113,6 +123,34 @@ describe("tools.fs.workspaceOnly", () => {
       ).rejects.toThrow(/Path escapes sandbox root/i);
       expect(await fs.readFile(path.join(agentRoot, "secret.txt"), "utf8")).toBe("shh");
     });
+  });
+
+  it("allows read-only agent workspace mounts for sandbox reads only", async () => {
+    await withUnsafeMountedSandboxHarness(
+      async ({ agentRoot, sandbox }) => {
+        await fs.writeFile(path.join(agentRoot, "secret.txt"), "shh", "utf8");
+
+        const tools = createSandboxFsTools({ sandbox, workspaceOnly: true });
+        const { readTool, writeTool, editTool } = expectReadWriteEditTools(tools);
+
+        const readResult = await readTool?.execute("t1", { path: "/agent/secret.txt" });
+        expect(getTextContent(readResult)).toContain("shh");
+
+        await expect(
+          writeTool?.execute("t2", { path: "/agent/owned.txt", content: "x" }),
+        ).rejects.toThrow(/Path escapes sandbox root/i);
+        const missingOwnedFile = await fs
+          .stat(path.join(agentRoot, "owned.txt"))
+          .catch((error: unknown) => error);
+        expect((missingOwnedFile as NodeJS.ErrnoException).code).toBe("ENOENT");
+
+        await expect(
+          editTool?.execute("t3", { path: "/agent/secret.txt", oldText: "shh", newText: "nope" }),
+        ).rejects.toThrow(/Path escapes sandbox root/i);
+        expect(await fs.readFile(path.join(agentRoot, "secret.txt"), "utf8")).toBe("shh");
+      },
+      { workspaceAccess: "ro" },
+    );
   });
 
   it("enforces apply_patch workspace-only in sandbox mounts by default", async () => {

--- a/src/agents/pi-tools.ts
+++ b/src/agents/pi-tools.ts
@@ -56,6 +56,7 @@ import {
 import { cleanToolSchemaForGemini, normalizeToolParameters } from "./pi-tools.schema.js";
 import type { AnyAgentTool } from "./pi-tools.types.js";
 import type { SandboxContext } from "./sandbox.js";
+import { SANDBOX_AGENT_WORKSPACE_MOUNT } from "./sandbox/constants.js";
 import {
   isSubagentEnvelopeSession,
   resolveSubagentCapabilityStore,
@@ -96,6 +97,29 @@ function isOpenAIProvider(provider?: string) {
 }
 
 const MEMORY_FLUSH_ALLOWED_TOOL_NAMES = new Set(["read", "write"]);
+
+type GuardContainerMount = {
+  containerRoot: string;
+  hostRoot: string;
+};
+
+function readOnlyAgentWorkspaceMount(
+  sandbox: SandboxContext | null | undefined,
+): GuardContainerMount[] | undefined {
+  if (
+    !sandbox ||
+    sandbox.workspaceAccess !== "ro" ||
+    sandbox.agentWorkspaceDir === sandbox.workspaceDir
+  ) {
+    return undefined;
+  }
+  return [
+    {
+      containerRoot: SANDBOX_AGENT_WORKSPACE_MOUNT,
+      hostRoot: sandbox.agentWorkspaceDir,
+    },
+  ];
+}
 
 type BashToolsModule = typeof import("./bash-tools.js");
 
@@ -592,6 +616,7 @@ export function createOpenClawCodingTools(options?: {
           base.push(
             workspaceOnly
               ? wrapToolWorkspaceRootGuardWithOptions(sandboxed, sandboxRoot, {
+                  additionalContainerMounts: readOnlyAgentWorkspaceMount(sandbox),
                   containerWorkdir: sandbox.containerWorkdir,
                 })
               : sandboxed,

--- a/src/agents/test-helpers/unsafe-mounted-sandbox.ts
+++ b/src/agents/test-helpers/unsafe-mounted-sandbox.ts
@@ -49,6 +49,7 @@ function createUnsafeMountedBridge(params: {
 export function createUnsafeMountedSandbox(params: {
   sandboxRoot: string;
   agentRoot: string;
+  workspaceAccess?: "none" | "ro" | "rw";
   workspaceContainerRoot?: string;
 }): SandboxContext {
   const bridge = createUnsafeMountedBridge({
@@ -59,7 +60,7 @@ export function createUnsafeMountedSandbox(params: {
   return createPiToolsSandboxContext({
     workspaceDir: params.sandboxRoot,
     agentWorkspaceDir: params.agentRoot,
-    workspaceAccess: "rw",
+    workspaceAccess: params.workspaceAccess ?? "rw",
     fsBridge: bridge,
     tools: { allow: [], deny: [] },
   });
@@ -67,13 +68,18 @@ export function createUnsafeMountedSandbox(params: {
 
 export async function withUnsafeMountedSandboxHarness(
   run: (ctx: { sandboxRoot: string; agentRoot: string; sandbox: SandboxContext }) => Promise<void>,
+  options?: { workspaceAccess?: "none" | "ro" | "rw" },
 ) {
   const stateDir = await fs.mkdtemp(path.join(os.tmpdir(), "openclaw-sbx-mounts-"));
   const sandboxRoot = path.join(stateDir, "sandbox");
   const agentRoot = path.join(stateDir, "agent");
   await fs.mkdir(sandboxRoot, { recursive: true });
   await fs.mkdir(agentRoot, { recursive: true });
-  const sandbox = createUnsafeMountedSandbox({ sandboxRoot, agentRoot });
+  const sandbox = createUnsafeMountedSandbox({
+    sandboxRoot,
+    agentRoot,
+    workspaceAccess: options?.workspaceAccess,
+  });
   try {
     await run({ sandboxRoot, agentRoot, sandbox });
   } finally {

